README.md
# improved_logging

Improves the log format and adds an optional `exception` parameter to the 
`warn()` and `error()` methods to print a stack trace automatically.

## Installation

```ruby
gem 'improved_logging'
```

In your `application.rb`, include the following (or similar):

```ruby
config.logger = ActiveSupport::BufferedLogger.new(config.paths.log.first)
config.logger.level = Logger::DEBUG
```

## Configuration

By default it behaves differently in development mode, it doesn't log the time, 
process ID or hostname. If `Rails.env` is not `development` those will be logged.

There are some options that can be set, add these lines to an initializer:

`ImprovedLogging.verbose = false`

 * This suppresses printing of the hostname, process ID and timestamp. This 
defaults to false in development mode, true otherwise.

`ImprovedLogging.hostname_maxlen = 15`

 * This sets the maximum number of characters of the hostname that will be 
printed. The beginning of the hostname is truncated, rather than the end, 
because often the end of the hostname is more unique than the beginning. 
The default is 10.

`ImprovedLogging.custom = "my-custom-string"`

 * This sets a custom string that will be printed at the beginning of the 
output, if desired.

## Format

It adds severity level (with colour, if colour is enabled), hostname and 
process ID to each log line, and it and adds warn() and exception() methods that 
take an exception as an argument and print the stack trace.

Sample verbose output:

<pre>
fossil.4685  Started GET "/my-mlg?page=1&page_size=10&_=1314650041431" for 127.0.0.1 at 2011-08-29 16:34:01 -0400
fossil.4685  16:34:01  INFO       Processing by TimelineEventsController#personal_and_friends as JSON
fossil.4685  16:34:01  INFO       Parameters: {"page"=>"1", "page_size"=>"10", "_"=>"1314650041431"}
fossil.4685  16:34:01  DEBUG      User Load (0.4ms)  SELECT `users`.* FROM `users` WHERE `users`.`nickname` = 'AgoraDavid' LIMIT 1
fossil.4685  16:34:01  DEBUG      CACHE (0.0ms)  SELECT `users`.* FROM `users` WHERE `users`.`nickname` = 'AgoraDavid' LIMIT 1
fossil.4685  16:34:01  DEBUG      TimelineEvent Load (0.5ms)  SELECT `timeline_events`.* FROM `timeline_events` WHERE (actor_id IN (1)) ORDER BY updated_at DESC LIMIT 10 OFFSET 0
fossil.4685  16:34:01  DEBUG      SQL (0.2ms)  SELECT COUNT(*) FROM `votes` WHERE `votes`.`voteable_id` = 1 AND `votes`.`voteable_type` = 'TimelineEvent' AND `votes`.`vote` = 0
fossil.4685  16:34:01  INFO     Rendered shared/event_message_templates/_user_created.html.haml (1.7ms)
fossil.4685  16:34:01  DEBUG      User Load (0.3ms)  SELECT `users`.* FROM `users` WHERE `users`.`id` = 1 LIMIT 1
fossil.4685  16:34:01  DEBUG      SQL (0.4ms)  SELECT COUNT(*) FROM `comments` WHERE (`comments`.commentable_id = 1 AND `comments`.commentable_type = 'TimelineEvent')
fossil.4685  16:34:01  DEBUG      Comment Load (0.2ms)  SELECT `comments`.* FROM `comments` WHERE (`comments`.commentable_id = 1 AND `comments`.commentable_type = 'TimelineEvent')
fossil.4685  16:34:01  DEBUG      CACHE (0.0ms)  SELECT COUNT(*) FROM `comments` WHERE (`comments`.commentable_id = 1 AND `comments`.commentable_type = 'TimelineEvent')
fossil.4685  16:34:01  INFO     Rendered shared/event_message_templates/_new_comment_form.html.haml (8.2ms)
fossil.4685  16:34:01  INFO     Rendered shared/event_message_templates/_comments.html.haml (22.9ms)
fossil.4685  16:34:01  INFO     Rendered shared/event_message_templates/_message_template.html.haml (69.2ms)
fossil.4685  16:34:01  INFO     Rendered shared/feed.haml (72.6ms)
fossil.4685  16:34:01  INFO     Completed 200 OK in 246ms (Views: 3.4ms | ActiveRecord: 2.0ms)
</pre>

## Contributing to improved_logging

 * Check out the latest master to make sure the feature hasn't been implemented or the bug hasn't been fixed yet
 * Check out the issue tracker to make sure someone already hasn't requested it and/or contributed it
 * Fork the project
 * Start a feature/bugfix branch
 * Commit and push until you are happy with your contribution
 * Make sure to add tests for it. This is important so I don't break it in a future version unintentionally.
 * Please try not to mess with the Rakefile, version, or history. If you want to have your own version, or is otherwise necessary, that is fine, but please isolate to its own commit so I can cherry-pick around it.

## License

Code originally found in https://github.com/pauldowman/better_logging

Uses the same license as better_logging, 
Creative Commons "Attribution-Share Alike", http://creativecommons.org/licenses/by-sa/3.0/


CHANGELOG.md
# 1.0.2

* Added support for yielding to message block if passed to warn or error methods

# 1.0.1

* Remove explicit dependency requirements
 
# 1.0.0

* Initial release

spec/spec_helper.rb
require 'rubygems'
require 'rspec'

RSpec.configure do |configuration|
end

spec/version_spec.rb
require 'spec_helper'
require 'improved_logging/version'

describe ImprovedLogging do
  it "should be the correct version" do
    ImprovedLogging::VERSION.should == '1.0.2'
  end
end

spec/improved_logging_spec.rb
require 'spec_helper'
require 'active_support/core_ext/module'
require 'active_support'
require 'improved_logging'

describe ActiveSupport::BufferedLogger do
  
  before(:each) do
    @logger = ActiveSupport::BufferedLogger.new(StringIO.new)
  end
  
  it 'should include the ImprovedLogging module' do
    ActiveSupport::BufferedLogger.included_modules.should include(::ImprovedLogging)
    @logger.methods.should include(:add_with_extra_info)
    @logger.methods.should include(:error_with_exception_param)
    @logger.methods.should include(:warn_with_exception_param)    
  end
  
  it 'should call #add when you log a message' do
    @logger.should_receive(:add).with(0, 'message', nil)
    
    @logger.debug('message')
  end
  
  it 'should allow you to pass an exception to the error method' do
    @logger.should_receive(:error).with('message', anything())
    
    begin
      raise StandardError.new
    rescue => e
      @logger.error('message', e)
    end
  end

  it 'should allow you to pass an exception to the warn method' do
    @logger.should_receive(:warn).with('message', anything())
    
    begin
      raise StandardError.new
    rescue => e
      @logger.warn('message', e)
    end
  end

  [:error, :warn].each do |severity|
    it 'should allow you to pass a block to the error method' do
      method_sym = "#{severity}_without_exception_param".to_sym
      @logger.should_receive(method_sym).with('message')
      begin
        raise StandardError.new
      rescue => e
        @logger.send(severity) { 'message' }
      end
    end
  end
end


lib/better_logging/better_logging.rb
# This module, when included into ActiveSupport::BufferedLogger, improves the
# logging format. See the README file for more info.
#
# This is distributed under a Creative Commons "Attribution-Share Alike"
# license: for details see: 
# http://creativecommons.org/licenses/by-sa/3.0/
#
module BetterLogging  
  LENGTH = ActiveSupport::BufferedLogger::Severity.constants.map{|c| c.to_s.length}.max
  
  def self.included(base)
    base.class_eval do
      alias_method_chain :add, :extra_info
      alias_method_chain :error, :exception_param
      alias_method_chain :warn, :exception_param
    end
    
    # Get the length to format the output so that the pid column lines up.
    # The severity levels probably won't change but this avoids hard-coding
    # them anyway, just in case.
    # Most of this is done with class_eval so it should only be done once 
    # while the class is being loaded.
    if_stmts = ""
    for c in ActiveSupport::BufferedLogger::Severity.constants
      if_stmts += <<-EOT
        if severity == #{c}
          severity_name = sprintf("%1$*2$s", "#{c}", #{LENGTH * -1})
          use_colour = false
          if Rails.version.to_i >= 3
            use_colour = true if ActiveSupport::LogSubscriber.colorize_logging
          else
            use_colour = true if defined?(ActiveRecord) && ActiveRecord::Base.colorize_logging
          end
          if use_colour
            if severity == INFO
              severity_name = "\033[32m" + severity_name + "\033[0m"
            elsif severity == WARN
              severity_name = "\033[33m" + severity_name + "\033[0m"
            elsif severity == ERROR || severity == FATAL
              severity_name = "\033[31m" + severity_name + "\033[0m"
            end
          end
          return severity_name
        end
      EOT
    end
    base.class_eval <<-EOT, __FILE__, __LINE__
      def self.severity_name(severity)
        #{if_stmts}
        return "UNKNOWN"
      end
    EOT
  end
  

  def self.verbose=(boolean)
    @@verbose = boolean
  end
  
  def self.custom=(string)
    @@custom = string
    @@line_prefix = format_line_prefix
  end
  
  def self.hostname_maxlen=(integer)
    @@hostname_maxlen = integer
    @@line_prefix = format_line_prefix
  end
  
  def self.format_line_prefix
    if @@full_hostname.length < @@hostname_maxlen
      hostname = @@full_hostname
    else
      hostname = @@full_hostname[-(@@hostname_maxlen)..-1]
    end
    
    line_prefix = sprintf("%1$*2$s", "#{hostname}.#{@@pid}  ", -(7 + hostname.length))
    line_prefix = "#{@@custom}  #{line_prefix}" if @@custom
    return line_prefix
  end
  
  def self.get_hostname
    `hostname -s`.strip
  end
  
  # The following are cached as class variables for speed.

  # These are configurable, put something like the following in an initializer:
  #   PaulDowman::RailsPlugins::BetterLogging.verbose = false
  @@verbose = Rails.env != "development"
  @@full_hostname = get_hostname
  @@hostname_maxlen = 10
  @@custom = nil
  
  # These are not configurable
  @@pid = $$
  @@line_prefix = format_line_prefix

        
  # the cached pid can be wrong after a fork(), this checks if it has changed and
  # re-caches the line_prefix
  def update_pid
    if @@pid != $$
      @@pid = $$
      @@line_prefix = BetterLogging.format_line_prefix
    end
  end
  
  def add_with_extra_info(severity, message = nil, progname = nil, &block)
    update_pid
    time = @@verbose ? "#{Time.new.strftime('%H:%M:%S')}  " : ""
    message = "#{time}#{ActiveSupport::BufferedLogger.severity_name(severity)}  #{message}"
    
    # Make sure every line has the PID and hostname and custom string 
    # so we can use grep to isolate output from one process or server.
    # gsub works even when the output contains "\n", though there's
    # probably a small performance cost.
    message = message.gsub(/^/, @@line_prefix) if @@verbose
    
    add_without_extra_info(severity, message, progname, &block)
  end
  
  
  # add an optional second parameter to the error & warn methods to allow a stack trace:
        
  def error_with_exception_param(message, exception = nil)
    message += "\n#{exception.inspect}\n#{exception.backtrace.join("\n")}" if exception
    error_without_exception_param(message)
  end
  
  def warn_with_exception_param(message, exception = nil)
    message += "\n#{exception.inspect}\n#{exception.backtrace.join("\n")}" if exception
    warn_without_exception_param(message)
  end
end


lib/improved_logging/improved_logging.rb
# This module, when included into ActiveSupport::BufferedLogger, improves the
# logging format. See the README file for more info.
#
# This is distributed under a Creative Commons "Attribution-Share Alike"
# license: for details see: 
# http://creativecommons.org/licenses/by-sa/3.0/
#
module ImprovedLogging  
  LENGTH = ::ActiveSupport::BufferedLogger::Severity.constants.map{|c| c.to_s.length}.max
  
  def self.included(base)
    base.class_eval do
      alias_method_chain :add, :extra_info
      alias_method_chain :error, :exception_param
      alias_method_chain :warn, :exception_param
    end
    
    # Get the length to format the output so that the pid column lines up.
    # The severity levels probably won't change but this avoids hard-coding
    # them anyway, just in case.
    # Most of this is done with class_eval so it should only be done once 
    # while the class is being loaded.
    if_stmts = ""
    for c in ::ActiveSupport::BufferedLogger::Severity.constants
      if_stmts += <<-EOT
        if severity == #{c}
          severity_name = sprintf("%1$*2$s", "#{c}", #{LENGTH * -1})
          use_colour = false
          if ::Rails.version.to_i >= 3
            use_colour = true if ::ActiveSupport::LogSubscriber.colorize_logging
          else
            use_colour = true if defined?(::ActiveRecord) && ::ActiveRecord::Base.colorize_logging
          end
          if use_colour
            if severity == INFO
              severity_name = "\033[32m" + severity_name + "\033[0m"
            elsif severity == WARN
              severity_name = "\033[33m" + severity_name + "\033[0m"
            elsif severity == ERROR || severity == FATAL
              severity_name = "\033[31m" + severity_name + "\033[0m"
            end
          end
          return severity_name
        end
      EOT
    end
    base.class_eval <<-EOT, __FILE__, __LINE__
      def self.severity_name(severity)
        #{if_stmts}
        return "UNKNOWN"
      end
    EOT
  end
  

  def self.verbose=(boolean)
    @@verbose = boolean
  end
  
  def self.custom=(string)
    @@custom = string
    @@line_prefix = format_line_prefix
  end
  
  def self.hostname_maxlen=(integer)
    @@hostname_maxlen = integer
    @@line_prefix = format_line_prefix
  end
  
  def self.format_line_prefix
    if @@full_hostname.length < @@hostname_maxlen
      hostname = @@full_hostname
    else
      hostname = @@full_hostname[-(@@hostname_maxlen)..-1]
    end
    
    line_prefix = sprintf("%1$*2$s", "#{hostname}.#{@@pid}  ", -(7 + hostname.length))
    line_prefix = "#{@@custom}  #{line_prefix}" if @@custom
    return line_prefix
  end
  
  def self.get_hostname
    `hostname -s`.strip
  end
  
  # The following are cached as class variables for speed.

  # These are configurable, put something like the following in an initializer:
  #   ImprovedLogging.verbose = false
  @@verbose = defined?(::Rails) ? ::Rails.env != "development" : false
  @@full_hostname = get_hostname
  @@hostname_maxlen = 10
  @@custom = nil
  
  # These are not configurable
  @@pid = $$
  @@line_prefix = format_line_prefix

        
  # the cached pid can be wrong after a fork(), this checks if it has changed and
  # re-caches the line_prefix
  def update_pid
    if @@pid != $$
      @@pid = $$
      @@line_prefix = ImprovedLogging.format_line_prefix
    end
  end
  
  def add_with_extra_info(severity, message = nil, progname = nil, &block)
    update_pid
    time = @@verbose ? "#{Time.new.strftime('%H:%M:%S')}  " : ""
    message = "#{time}#{::ActiveSupport::BufferedLogger.severity_name(severity)}  #{message}"
    
    # Make sure every line has the PID and hostname and custom string 
    # so we can use grep to isolate output from one process or server.
    # gsub works even when the output contains "\n", though there's
    # probably a small performance cost.
    message = message.gsub(/^/, @@line_prefix) if @@verbose
    
    add_without_extra_info(severity, message, progname, &block)
  end
  
  
  # add an optional second parameter to the error & warn methods to allow a stack trace:
        
  def error_with_exception_param(message = '', exception = nil)
    message = yield if block_given?
    message += "\n#{exception.inspect}\n#{exception.backtrace.join("\n")}" if exception
    error_without_exception_param(message)
  end
  
  def warn_with_exception_param(message = '', exception = nil)
    message = yield if block_given?
    message += "\n#{exception.inspect}\n#{exception.backtrace.join("\n")}" if exception
    warn_without_exception_param(message)
  end
end
